File: SortVision/public/code/bubble/csharp/bubbleSort.cs
using System;

namespace SortVision
{
    /// <summary>
    /// A class that provides Bubble Sort functionality.
    /// </summary>
    public class BubbleSort
    {
        /// <summary>
        /// Sorts an integer array using the Bubble Sort algorithm.
        /// </summary>
        /// <param name="arr">The array to be sorted.</param>
        public static void Sort(int[] arr)
        {
            if (arr == null || arr.Length <= 1)
            {
                // Array is null, empty, or has a single element; no sorting needed. d
                return;
            }

            int n = arr.Length;
            for (int i = 0; i < n - 1; i++)
            {
                // Track if any swapping is done in this pass
                bool swapped = false;

                for (int j = 0; j < n - i - 1; j++)
                {
                    // Compare adjacent elements
                    if (arr[j] > arr[j + 1])
                    {
                        // Swap if needed
                        int temp = arr[j];
                        arr[j] = arr[j + 1];
                        arr[j + 1] = temp;

                        swapped = true;
                    }
                }

                // If no swaps occurred, the array is already sorted
                if (!swapped)
                {
                    break;
                }
            }
        }

        /// <summary>
        /// Main method for testing the BubbleSort implementation.
        /// </summary>
        public static void Main(string[] args)
        {
            // Test cases
            int[][] testCases = {
                new int[] { 64, 34, 25, 12, 22, 11, 90 }, // Standard case
                new int[] { 1 },                          // Single element
                new int[] { },                            // Empty array
                new int[] { 1, 2, 3, 4, 5 },              // Already sorted
                new int[] { 5, 4, 3, 2, 1 },              // Reverse sorted
                new int[] { 3, 3, 1, 2, 3 }               // Contains duplicates
            };

            Console.WriteLine("Bubble Sort Test Cases:");
            foreach (var testCase in testCases)
            {
                Console.WriteLine("\nOriginal Array: " + string.Join(", ", testCase));
                Sort(testCase);
                Console.WriteLine("Sorted Array:   " + string.Join(", ", testCase));
            }
        }
    }
}

File: SortVision/public/code/bubble/kotlin/bubbleSort.kt
package com.sortvision

import java.util.Arrays

/**
 * Implements the Bubble Sort algorithm.
 *
 * Time Complexity:
 * - Best Case: O(n) - Occurs when the array is already sorted.
 *   The algorithm makes one pass through the array to confirm it's sorted, thanks to the `swapped` flag optimization.
 * - Average Case: O(n^2) - Occurs when the elements are in a jumbled order.
 * - Worst Case: O(n^2) - Occurs when the array is sorted in reverse order.
 *
 * Space Complexity: O(1) - Bubble sort is an in-place sorting algorithm.
 *   It requires a constant amount of extra space for temporary variables (e.g., for swapping elements).
 *   The space used by the input array itself is not counted in this O(1) complexity.
 */
class BubbleSort {
    /**
     * Sorts an array of integers using the Bubble Sort algorithm.
     *
     * @param arr The array to be sorted.
     * @return The sorted array.
     */
    fun sort(arr: IntArray): IntArray {
        // Handle edge cases for empty or single-element array
        if (arr.size <= 1) {
            return arr
        }

        val n = arr.size
        var swapped: Boolean
        // Outer loop for passes
        for (i in 0 until n - 1) {
            swapped = false
            // Inner loop for comparisons and swaps
            for (j in 0 until n - i - 1) {
                // If the element found is greater than the next element, swap them
                if (arr[j] > arr[j + 1]) {
                    val temp = arr[j]
                    arr[j] = arr[j + 1]
                    arr[j + 1] = temp
                    swapped = true // Mark that a swap occurred in this pass
                }
            }
            // Optimization: If no two elements were swapped by inner loop in this pass,
            // then the array is already sorted, and we can break early.
            if (!swapped) {
                break
            }
        }
        return arr
    }
}

fun main() {
    val sorter = BubbleSort()
    var testArray: IntArray
    var sortedArray: IntArray

    // Test case 1: Empty array
    testArray = intArrayOf()
    println("Original array: ${Arrays.toString(testArray)}")
    sortedArray = sorter.sort(testArray.copyOf())
    println("Sorted array: ${Arrays.toString(sortedArray)}")
    println("---")

    // Test case 2: Single-element array
    testArray = intArrayOf(5)
    println("Original array: ${Arrays.toString(testArray)}")
    sortedArray = sorter.sort(testArray.copyOf())
    println("Sorted array: ${Arrays.toString(sortedArray)}")
    println("---")

    // Test case 3: Already sorted array
    testArray = intArrayOf(1, 2, 3, 4, 5)
    println("Original array: ${Arrays.toString(testArray)}")
    sortedArray = sorter.sort(testArray.copyOf())
    println("Sorted array: ${Arrays.toString(sortedArray)}")
    println("---")

    // Test case 4: Reverse-sorted array
    testArray = intArrayOf(5, 4, 3, 2, 1)
    println("Original array: ${Arrays.toString(testArray)}")
    sortedArray = sorter.sort(testArray.copyOf())
    println("Sorted array: ${Arrays.toString(sortedArray)}")
    println("---")

    // Test case 5: Array with duplicate elements
    testArray = intArrayOf(3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5)
    println("Original array: ${Arrays.toString(testArray)}")
    sortedArray = sorter.sort(testArray.copyOf())
    println("Sorted array: ${Arrays.toString(sortedArray)}")
    println("---")

    // Test case 6: Array with negative numbers
    testArray = intArrayOf(-5, -2, -8, -1, -4)
    println("Original array: ${Arrays.toString(testArray)}")
    sortedArray = sorter.sort(testArray.copyOf())
    println("Sorted array: ${Arrays.toString(sortedArray)}")
    println("---")

    // Test case 7: Mixed array of positive and negative numbers
    testArray = intArrayOf(0, 5, -2, 8, -1, 4, -4)
    println("Original array: ${Arrays.toString(testArray)}")
    sortedArray = sorter.sort(testArray.copyOf())
    println("Sorted array: ${Arrays.toString(sortedArray)}")
    println("---")
}
